cmd/kaspawallet/daemon/server/split_transaction.go

```diff
@@ -0,0 +1,258 @@
+		// 1. For the payment itself
+		// 2. (optional) for change
+		return nil, errors.Errorf("original transaction has %d outputs, while 1 or 2 are expected",
+			len(originalTransaction.Tx.Outputs))
+	}
+
+	totalValue := uint64(0)
+	sentValue := originalTransaction.Tx.Outputs[0].Value
+	utxos := make([]*libkaspawallet.UTXO, len(splitTransactions))
+	for i, splitTransaction := range splitTransactions {
+		output := splitTransaction.Tx.Outputs[0]
+		utxos[i] = &libkaspawallet.UTXO{
+			Outpoint: &externalapi.DomainOutpoint{
+				TransactionID: *consensushashing.TransactionID(splitTransaction.Tx),
+				Index:         0,
+			},
+			UTXOEntry:      utxo.NewUTXOEntry(output.Value, output.ScriptPublicKey, false, constants.UnacceptedDAAScore),
+			DerivationPath: s.walletAddressPath(changeWalletAddress),
+		}
+		totalValue += output.Value
+		totalValue -= feePerInput
+	}
+
+	if totalValue < sentValue {
+		// sometimes the fees from compound transactions make the total output higher than what's available from selected
+		// utxos, in such cases - find one more UTXO and use it.
+		additionalUTXOs, totalValueAdded, err := s.moreUTXOsForMergeTransaction(utxos, sentValue-totalValue)
+		if err != nil {
+			return nil, err
+		}
+		utxos = append(utxos, additionalUTXOs...)
+		totalValue += totalValueAdded
+	}
+
+	payments := []*libkaspawallet.Payment{{
+		Address: toAddress,
+		Amount:  sentValue,
+	}}
+	if totalValue > sentValue {
+		payments = append(payments, &libkaspawallet.Payment{
+			Address: changeAddress,
+			Amount:  totalValue - sentValue,
+		})
+	}
+
+	mergeTransactionBytes, err := libkaspawallet.CreateUnsignedTransaction(s.keysFile.ExtendedPublicKeys,
+		s.keysFile.MinimumSignatures, payments, utxos)
+	if err != nil {
+		return nil, err
+	}
+
+	return serialization.DeserializePartiallySignedTransaction(mergeTransactionBytes)
+}
+
+func (s *server) maybeSplitTransaction(transaction *serialization.PartiallySignedTransaction,
+	changeAddress util.Address) ([]*serialization.PartiallySignedTransaction, error) {
+
+	transactionMass, err := s.estimateMassAfterSignatures(transaction)
+	if err != nil {
+		return nil, err
+	}
+
+	if transactionMass < mempool.MaximumStandardTransactionMass {
+		return []*serialization.PartiallySignedTransaction{transaction}, nil
+	}
+
+	splitCount, inputCountPerSplit := s.splitAndInputPerSplitCounts(transaction, transactionMass)
+
+	splitTransactions := make([]*serialization.PartiallySignedTransaction, splitCount)
+	for i := 0; i < splitCount; i++ {
+		startIndex := i * inputCountPerSplit
+		endIndex := startIndex + inputCountPerSplit
+		var err error
+		splitTransactions[i], err = s.createSplitTransaction(transaction, changeAddress, startIndex, endIndex)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return splitTransactions, nil
+}
+
+func (s *server) splitAndInputPerSplitCounts(transaction *serialization.PartiallySignedTransaction, transactionMass uint64) (
```

Comment:
For the sake of argument, let's assume our transaction contains:
* 9 pay to public key inputs, each one has mass 100
* 1 multisig input with mass 900
* Mass limit of 1000
* A transaction without any inputs has mass 50

I believe this method would produce:
* massWithoutInputs = 50
* massForInputs = 1800
* inputCount = 10
* massPerInput = 180
* inputPerSplitCount = 5
* splitCount = 2

The splitCount is correct, but if we try to have inputPerSplitCount of 5, we'd get:
* One transaction with mass 550
* Another transaction with mass **1350**

1350 is over the mass limit of 1000

Reply:
We assume that the transaction was created by kaspawallet, thus it wouldn't have inputs with different masses.
Therefore your scenario is impossible.

Reply:
I've added a comment.
Adding a sanity test means either refactoring the CalculateMass code to have CalculateMassForInput, or using some not-entirely-accurate proxy such as input size, and I'd rather not do either.

Reply:
You can check if all inputs are multisig with the same number of required signatures, or all of them are p2pk